### PR TITLE
Proposal of selector generation

### PR DESCRIPTION
**Pull request by mengqiy**
Selector generation for Deployments, ReplicaSet, DaemonSet, StatefulSet and ReplicationController.

Originally discussed in #187.

cc: @pwittrock @erictune

### CODE REVIEW

**Comment by enisoc** [comment]
The [controllerRef proposal](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/controller-ref.md) already claims to solve the "selector overlap" problem, and it was chosen over generated labels for that purpose because of the points discussed here: https://github.com/kubernetes/kubernetes/issues/36859#issuecomment-262022027.

I'm working on finishing the [implementation of controllerRef](https://github.com/kubernetes/kubernetes/issues/24946). If that works to solve selector overlap, is there an additional benefit to expanding selector generation to all the workload controllers?

**Comment by kargakis** [comment]
> The controllerRef proposal already claims to solve the "selector overlap" problem, and it was chosen over generated labels for that purpose because of the points discussed here: kubernetes/kubernetes#36859 (comment).
> 
> I'm working on finishing the implementation of controllerRef. If that works to solve selector overlap, is there an additional benefit to expanding selector generation to all the workload controllers?

Owner references don't fully solve selector overlap. Between two overlapping deployments A and B (assuming A is older thus works as expected), B cannot adopt the replica sets owned by A but B is still blocked. A unique selector key would address that issue.

**Comment by kargakis** [comment]
> Owner references don't fully solve selector overlap. Between two overlapping deployments A and B (assuming A is older thus works as expected), B cannot adopt the replica sets owned by A but B is still blocked. A unique selector key would address that issue.

And even if this worked, we could end up in weird situations:
Deployment A selects RS-A1, RS-A2, RS-A3
Deployment B selects RS-B1, RS-B2

Deployment A is deleted.
The garbage collector now races with Deployment B, either it will get to the replica sets first and delete them or Deployment B adopts the additional replica sets and ends up with RS-B1, RS-B2, RS-A1, RS-A2, RS-A3 in the worst case.

File: contributors/design-proposals/selector-generation-for-all.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+The [Selector Generation](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/selector-generation.md) works well in Job to make sure non-overlapping. We can use this pattern for Deployments, ReplicaSet, DaemonSet, StatefulSet and ReplicationController.
+If `spec.selector` is not derived from `spec.template.metadata.labels`, `kubectl apply` will work well for selectors.
+
+## API
+
+The change will be similar to [the changes to Job](https://github.com/kubernetes/kubernetes/blob/master/docs/design/selector-generation.md#api). But the defaulting is different from Job's.
+
+`extension/v1beta1 Deployment|ReplicaSet|DaemonSet`, `apps/v1beta1 StatefulSet` and `v1 ReplicationController` change as follows.
+
+Add field `spec.manualSelector`. It controls if using automatic generated selectors.
```
Comment (kargakis) [inline contributors/design-proposals/selector-generation-for-all.md]: I am not sure I like the additional field. I suggest we always default a key - is there a backwards-compatibility concern here?

Comment (kargakis) [inline contributors/design-proposals/selector-generation-for-all.md]: Users already hardly understand how to use selectors, adding one more field to work-around that will probably cause more confusion than less.

**Comment by kargakis** [comment]
> Owner references don't fully solve selector overlap. Between two overlapping deployments A and B (assuming A is older thus works as expected), B cannot adopt the replica sets owned by A but B is still blocked. A unique selector key would address that issue.

Thinking about it more, if we change our listers to take into account owner references, we may not need selector generation. Haven't looked closely at that but it sounds plausible.

**Comment by enisoc** [comment]
> Thinking about it more, if we change our listers to take into account owner references, we may not need selector generation. Haven't looked closely at that but it sounds plausible.

I had originally thought this was included in the [controllerRef proposal](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/controller-ref.md), but upon re-reading I see that it's not quite clear what the intention was. The question specifically is what the implications of "ownership" are in this passage:

> Most controllers (all that manage collections of things defined by label selector) will have slightly changed semantics: currently controller owns an object if its selector matches object’s labels and if it doesn't notice an older controller of the same kind that also matches the object's labels, but after introduction of ControllerReference a controller will own an object iff selector matches labels and the OwnerReference with Controller=true points to it.

If the controller ignores an object because its controllerRef points to someone else, I thought that means it will create its own object, so no controller will get stuck. I'll try to get clarification from the reviewer of that proposal.

**Comment by lavalamp** [comment]
> Owner references don't fully solve selector overlap. Between two overlapping deployments A and B (assuming A is older thus works as expected), B cannot adopt the replica sets owned by A but B is still blocked. A unique selector key would address that issue.

That's not how it works :)

B is not blocked-- if A wants 10 pod and B wants 10 pods, you'll get total 20 pods. A and B will each construct their pods with the correct controllerRef already specified, so there will never be any un-adopted pods and no opportunity for the two controllers to fight over anything.

File: contributors/design-proposals/selector-generation-for-all.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
+Selector Generation for Deployments, ReplicaSet, DaemonSet, StatefulSet and ReplicationController
+=============
+
+# Goals
+Make selector easy to use and less error-prone for Deployments, ReplicaSet, DaemonSet, StatefulSet and ReplicationController. Make `kubectl apply` work well with selector.
```
Comment (lavalamp) [inline contributors/design-proposals/selector-generation-for-all.md]: Can we remove all wording in this doc that's about correctness? `controllerRef` should take care of the existing problems. I think there is something that needs to be done to fix `kubectl apply`, but this is a really invasive change and we should think about whether anything else can be done.

File: contributors/design-proposals/selector-generation-for-all.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+# Problem Description
+The field `spec.selector` of Deployments, ReplicaSet, DaemonSet are defaulted from `spec.template.metadata.labels`, if it is unspecified.
+And there is validation to make sure `spec.selector` always selects the pod template.
+The defaulting of selector may prevent from `kubectl apply` from working when updating the `spec.selector` field. e.g. [kubernetes/kubernetes#26202 (comment)](https://github.com/kubernetes/kubernetes/issues/26202#issuecomment-221421254)
```
Comment (lavalamp) [inline contributors/design-proposals/selector-generation-for-all.md]: Can we add an "alternatives considered" section to this doc? I would at least like to see what it would look like to fix this problem by e.g. modifying `kubectl apply` instead of every other component :)

File: contributors/design-proposals/selector-generation-for-all.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+If user does specify `spec.selector` then the user must also specify `spec.manualSelector`. This ensures the user knows that what he is doing is not the normal thing to do.
+
+## Controllers
+
+Controller managers are responsible to set generated labels when adopting.
+
+## Garbage collector
+
+Garbage collector is responsible for cleaning up the extra generated labels when orphaning.
```
Comment (lavalamp) [inline contributors/design-proposals/selector-generation-for-all.md]: Why would this be necessary? I would prefer for GC to not have to do this.

File: contributors/design-proposals/selector-generation-for-all.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+### Rationale
+
+Label combination of `resource name` and `resource kind` is unique across all kind of resources.
+They are more predictable and human-friendly than UID.
+
+If user does specify `spec.selector` then the user must also specify `spec.manualSelector`. This ensures the user knows that what he is doing is not the normal thing to do.
+
+## Controllers
+
+Controller managers are responsible to set generated labels when adopting.
```
Comment (lavalamp) [inline contributors/design-proposals/selector-generation-for-all.md]: This doesn't make a lot of sense to me, controllers wouldn't adopt anything that didn't already have the generated label set, no?

File: contributors/design-proposals/selector-generation-for-all.md [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  - The first label is `resource-kind=$KINDOFRESOURCE`.
+  - The second label is `resource-name=$NAMEOFRESOURCE`.
+
+### Manual Mode
+Manual mode is the same as [Job's](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/selector-generation.md#manual-mode).
+
+### Rationale
+
+Label combination of `resource name` and `resource kind` is unique across all kind of resources.
+They are more predictable and human-friendly than UID.
```
Comment (lavalamp) [inline contributors/design-proposals/selector-generation-for-all.md]: It is not actually unique unless you include the group... I would really recommend against doing this.

File: contributors/design-proposals/selector-generation-for-all.md (hunk 4 above)
Comment (lavalamp) [inline contributors/design-proposals/selector-generation-for-all.md]: Yeah, I really really don't want to build knowledge of various labels into the GC's orphaning process.

**Comment by ymqytw** [comment]
Updated the proposal:
- propose a  client-side design
- move the original server-side design to `Alternatives considered` section.

**Comment by ymqytw** [comment]
Updated the proposal:

- propose selector generation as a plug-in in admission controller

**Comment by bgrant0607** [comment]
Please do not merge this without my approval.

File: contributors/design-proposals/selector-generation-from-kubectl.md [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+  - `matchLabels["resource-name"]="$NAME"`
+- Append 3 labels to `spec.template.metadata.labels`
+  - `resource-group=$GROUP`
+  - `resource-kind=$KIND`
+  - `resource-name=$NAME`
+
+The second set of labels is:
+
+- Add 1 more label, which concatenates all the three in the first set, to `spec.selector`
+  - `matchLabels["controllers.k8s.io/selector"]="$GROUP/$KIND/$NAME"`
```
Comment (bgrant0607) [inline contributors/design-proposals/selector-generation-from-kubectl.md]: Why is this needed if we have the other 3 labels?

Also, I try to avoid label keys with prefixes.

File: contributors/design-proposals/selector-generation-from-kubectl.md [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+- Append 1 label to `spec.template.metadata.labels`
+  - `controllers.k8s.io/selector=$GROUP/$KIND/$NAME`
+
+The third set of labels is:
+
+- Add 1 more label to `spec.selector`
+  - `matchLabels["resource-uid"]="$UIDOFRESOURCE"`
+- Append 1 label to `spec.template.metadata.labels`
+  - `resource-uid=$UIDOFRESOURCE`
```
Comment (bgrant0607) [inline contributors/design-proposals/selector-generation-from-kubectl.md]: Don't see why uid is needed.

File: contributors/design-proposals/selector-generation-from-kubectl.md [hunk 9]
```diff
@@ -0,0 +1,174 @@
+Selector generation from kubectl
+==================
+
+# Goals
+
+Make `kubectl apply` work well with selector.
```
Comment (soltysh) [inline contributors/design-proposals/selector-generation-from-kubectl.md]: The selector generation has to be done on server, and that should be the goal.

Comment (kargakis) [inline contributors/design-proposals/selector-generation-from-kubectl.md]: Agreed with @soltysh. Also I believe selector generation is going to be useful for more than just making `kubectl apply` work. For example, in Deployments we already need to generate a label for making the underlying ReplicaSets non-overlapping.

Comment (soltysh) [inline contributors/design-proposals/selector-generation-from-kubectl.md]: I'm inclined to say that everything should be using generated selectors, by default. With the ability to turn it off by a user, when needed. Selectors are for controllers to work with underlying resources, usually. There's nothing stooping users from adding additional labels, if needed.

Comment (kargakis) [inline contributors/design-proposals/selector-generation-from-kubectl.md]: So I skimmed through the proposal and it seems to target the `apply` use case in the client. It will most likely be superseded if `apply` is moved server-side. I need to read the selector generation proposal for Jobs.

**Comment by bgrant0607** [comment]
This approach isn't aligned with the intent of labels, so I'm closing this proposal.

cc @kow3ns
